File: clojure/strain/src/strain.clj
(ns strain)

(defn retain [f l] 
  (for 
    [
     el l
    :when (f el)
     ]
    el))

(defn discard [f l]
  (for 
    [
     el l
    :when (not (f el))
     ]
    el)
)

File: scala/strain/src/main/scala/Strain.scala
object Strain {
  def keep[T](list: Seq[T], predicate: (T) => Boolean): Seq[T] = {
    for(element <- list if predicate(element))
    yield element
  }
  def discard[T](list: Seq[T], predicate: (T) => Boolean): Seq[T] = {
    for(element <- list if !predicate(element))
    yield element
  }
}